code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPropositionItem.java

```diff
@@ -113,82 +100,137 @@ void track(final PropositionEventType interaction) {
+        } catch (final DataReaderException dataReaderException) {
+            Log.trace(LOG_TAG, SELF_TAG, "Exception occurred creating MessagingPropositionItem from rule consequence: %s", dataReaderException.getLocalizedMessage());
         }
-        return inboundContent;
+
+        return propositionItem;
+    }
+
+    /**
+     * Returns this {@link MessagingPropositionItem}'s content as a json content {@code Map<String, Object>}.
+     *
+     * @return {@link Map<String, Object>} object containing the {@link MessagingPropositionItem}'s content.
+     */
+    public Map<String, Object> getJsonContentMap() {
+        final JsonContentSchemaData schemaData = createSchemaData(SchemaType.JSON_CONTENT);
+        return schemaData != null ? schemaData.getJsonObjectContent() : null;
+    }
+
+    /**
+     * Returns this {@link MessagingPropositionItem}'s content as a json content {@code List<Map<String, Object>>}.
+     *
+     * @return {@link List<Map<String, Object>>} object containing the {@link MessagingPropositionItem}'s content.
+     */
+    public List<Map<String, Object>> getJsonArrayList() {
+        final JsonContentSchemaData schemaData = createSchemaData(SchemaType.JSON_CONTENT);
+        return schemaData != null ? schemaData.getJsonArrayContent() : null;
+    }
+
+    /**
+     * Returns this {@link MessagingPropositionItem}'s content as a html content {@code String}.
+     *
+     * @return {@link String} containing the {@link MessagingPropositionItem}'s content.
+     */
+    public String getHtmlContent() {
+        final HtmlContentSchemaData schemaData = createSchemaData(SchemaType.HTML_CONTENT);
+        return schemaData != null ? schemaData.getContent().toString() : null;
+    }
+
+    /**
+     * Returns this {@link MessagingPropositionItem}'s content as a {@code InAppSchemaData} object.
+     *
+     * @return {@link InAppSchemaData} object containing the {@link MessagingPropositionItem}'s content.
+     */
+    public InAppSchemaData getInAppSchemaData() {
+        return createSchemaData(SchemaType.INAPP);
+    }
+
+    /**
+     * Returns this {@link MessagingPropositionItem}'s content as a {@code FeedItemSchemaData} object.
+     *
+     * @return {@link FeedItemSchemaData} object containing the {@link MessagingPropositionItem}'s content.
+     */
+    public FeedItemSchemaData getFeedItemSchemaData() {
+        return createSchemaData(SchemaType.FEED);
     }
 
     /**
-     * Creates a {@code PropositionItem} object from the provided {@code Map<String, Object>}.
+     * Creates a schema data object from this {@code MessagingPropositionItem}'s content.
+     *
+     * @param schemaType {@link SchemaType} to be used when creating the {@link T} object.
+     * @return {@code T} object created from the provided {@link MessagingPropositionItem}'s content.
+     */
+    private <T> T createSchemaData(final SchemaType schemaType) {
```

Comment:
I would recommend having all schema data types extend from a common interface/abstract class. Would make the code more extensible to new types in the future

Reply:
We can use Curiously Recurring Template Pattern here, similar to what Core has done with Presentation types in the new Jetpack Compose refactor
```
public abstract class SchemaData<T extends SchemaData<T>>{
                abstract Object getContent();
}

public class HtmlContentSchemaData extends SchemaData<HtmlContentSchemaData> {
             public Object getContent() {
                      //HTML specific implementation for getContent
             }
}


public class FeedItemSchemaData extends SchemaData<FeedItemSchemaData> {
             public Object getContent() {
                      //Feed specific implementation for getContent
             }
}

public class InAppSchemaData extends SchemaData<InAppSchemaData> {
             public Object getContent() {
                      //Feed specific implementation for getContent
             }
}

private <T> T createSchemaData(final SchemaType schemaType) {
           // all switch cases, no need to cast
            case HTML_CONTENT:
                      return (T) new HtmlContentSchemaData(ruleJson);
}
```

Reply:
good point, i'll make this change